File: problems/CSES/Dynamic_Programming/rectangle_cutting/rectangle_cutting.py
import sys

input = sys.stdin.readline

def main():
    h, w = map(int, input().split())
    n = max(h, w)

    dp = [[0] * (n + 1) for _ in range(n + 1)]

    for b in range(1, n + 1):
        for a in range(1, b + 1):
            if a == b:
                continue

            best = a * b - 1

            for cut in range(1, a // 2 + 1):
                candidate = (
                    1
                    + dp[cut][b]
                    + dp[a - cut][b]
                )

                if candidate < best:
                    best = candidate

            for cut in range(1, b // 2 + 1):
                first_a = min(a, cut)
                first_b = max(a, cut)

                second = b - cut
                second_a = min(a, second)
                second_b = max(a, second)

                candidate = (
                    1
                    + dp[first_a][first_b]
                    + dp[second_a][second_b]
                )

                if candidate < best:
                    best = candidate

            dp[a][b] = best
            dp[b][a] = best

    print(dp[h][w])

main()

File: problems/CSES/Dynamic_Programming/rectangle_cutting/rectangle_cutting.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int h, w;
    cin >> h >> w;

    int n = max(h, w);

    vector<vector<int>> dp(
        n + 1,
        vector<int>(n + 1, 0)
    );

    for (int b = 1; b <= n; ++b) {
        for (int a = 1; a <= b; ++a) {
            if (a == b) {
                continue;
            }

            int best = a * b - 1;

            for (int cut = 1; cut <= a / 2; ++cut) {
                int candidate =
                    1 +
                    dp[cut][b] +
                    dp[a - cut][b];

                best = min(best, candidate);
            }

            for (int cut = 1; cut <= b / 2; ++cut) {
                int firstA = min(a, cut);
                int firstB = max(a, cut);

                int second = b - cut;
                int secondA = min(a, second);
                int secondB = max(a, second);

                int candidate =
                    1 +
                    dp[firstA][firstB] +
                    dp[secondA][secondB];

                best = min(best, candidate);
            }

            dp[a][b] = best;
            dp[b][a] = best;
        }
    }

    cout << dp[h][w] << '\n';

    return 0;
}
